src/core/contracts/utilities.ts

```diff
@@ -181,3 +185,148 @@ export const subscribeToEvent = async (
+
+/**
+ *  AsyncPublicationsIterator is an AsyncIterator which implements only an async next() function.
+ *  Lazily fetches log events from the chain based on a Publications filter, walkbackBlocks
+ *  and a range of blocks provided in the constructor.  When logs are found, they are
+ *  decoded into Publications.
+ */
+export class AsyncPublicationsIterator {
+  currentStartBlock: number;
+  currentEndBlock: number;
+  logIndex: number;
+  publications: Array<BatchPublicationLogData>;
+  provider: ethers.providers.Provider;
+  blockRangeOptions: BlockRangeOptions;
+
+  /**
+   *
+   * @param rangeOptions - BlockRangeOptions for fetching publications
+   * @param provider - ethers.providers.Provider
+   */
+  constructor(rangeOptions: BlockRangeOptions, provider: ethers.providers.Provider) {
+    this.blockRangeOptions = rangeOptions;
+    this.currentEndBlock = rangeOptions.latestBlock;
+    this.currentStartBlock = rangeOptions.latestBlock - rangeOptions.walkbackBlocks + 1;
+    this.provider = provider;
+    this.logIndex = 0;
+    this.publications = [];
+  }
+
+  reachedEarliestBlock(): boolean {
+    return this.currentEndBlock < this.blockRangeOptions.earliestBlock;
+  }
+
+  // we should fetch logs from the chain initially or if we've returned the last item
+  // fetched last time.
+  shouldFetchLogs(): boolean {
+    return !this.publications.length || this.logIndex === this.publications.length - 1;
+  }
+
+  doneFetching(): boolean {
+    return this.reachedEarliestBlock() && this.logIndex === this.publications.length;
+  }
+
+  async fetchUntilGetLogsOrDone(): Promise<void> {
+    let logs: EventLog[] = [];
+    // keep going until we get something or we hit the earliest requested block height.
+    while (!logs.length && !this.reachedEarliestBlock()) {
+      logs = await this.provider.getLogs({
+        address: this.blockRangeOptions.filter.address,
+        topics: this.blockRangeOptions.filter.topics,
+        fromBlock: this.currentStartBlock,
+        toBlock: this.currentEndBlock,
+      });
+      this.currentEndBlock = this.currentStartBlock - 1;
+      this.currentStartBlock = Math.max(
+        this.currentStartBlock - this.blockRangeOptions.walkbackBlocks,
+        this.blockRangeOptions.earliestBlock
+      );
+    }
+    this.logIndex = 0;
+    this.publications = decodeLogsForBatchPublication(logs);
+    return;
+  }
+
+  /**
+   * next - returns a Promise of an IteratorResult enclosing a Publication.
+   * IteratorResult = \{
+   *    done: true if the earliest block has been reached AND the last Publication has
+   *          been returned via next().
+   *    value: a Publication
+   * \}
+   */
+  public async next(): Promise<IteratorResult<BatchPublicationLogData>> {
+    if (this.shouldFetchLogs()) {
+      await this.fetchUntilGetLogsOrDone();
+    } else {
+      this.logIndex++;
+    }
+    return Promise.resolve({ done: this.doneFetching(), value: this.publications[this.logIndex] });
+  }
```

Comment:
We need this to support `for await (const nextResult of iterator)`

```suggestion
  }
  
  public [Symbol.asyncIterator](): AsyncIterableIterator<BatchPublicationLogData> {
    return this;
  }
```